persist/fs/commitlog/reader.go

```diff
@@ -109,58 +158,166 @@ func (r *reader) Read() (
+		}
+		decoder.Reset(arg.bytes.Get())
+		entry, err := decoder.DecodeLogEntry()
+		if err != nil {
+			readResponse.resultErr = err
+			outBuf <- readResponse
+			continue
 		}
-		ok = true
 
-		r.metadataLookup[entry.Index] = metadata
-	}
+		if len(entry.Metadata) != 0 {
+			metadataDecoder.Reset(entry.Metadata)
+			decoded, err := metadataDecoder.DecodeLogMetadata()
+			if err != nil {
+				readResponse.resultErr = err
+				outBuf <- readResponse
+				continue
+			}
 
-	if !ok {
-		resultErr = errCommitLogReaderMissingLogMetadata
-		return
-	}
+			id := r.bytesPool.Get(len(decoded.ID))
+			id.IncRef()
+			id.AppendAll(decoded.ID)
+
+			namespace := r.bytesPool.Get(len(decoded.Namespace))
+			namespace.IncRef()
+			namespace.AppendAll(decoded.Namespace)
+
+			r.metadataLock.Lock()
+			r.metadataLookup[entry.Index] = Series{
+				UniqueIndex: entry.Index,
+				ID:          ts.BinaryID(id),
+				Namespace:   ts.BinaryID(namespace),
+				Shard:       decoded.Shard,
+			}
+			waitGroup, ok := r.metadataWaitGroups[entry.Index]
+			// Another goroutine is blocked waiting for this metadata
+			if ok {
+				waitGroup.Done()
+			}
+			r.metadataLock.Unlock()
+			namespace.DecRef()
+			id.DecRef()
+		}
+
+		var pendingResult *sync.WaitGroup
+		var metadata Series
+		var hasMetadata bool
+		var ok bool
+
+		// Try cheap read lock first
+		r.metadataLock.RLock()
+		metadata, hasMetadata = r.metadataLookup[entry.Index]
+		r.metadataLock.RUnlock()
+
+		// The required metadata hasn't been processed yet
+		r.metadataLock.Lock()
+		if !hasMetadata {
+			metadata, hasMetadata = r.metadataLookup[entry.Index]
+			if !hasMetadata {
+				waitGroup, hasWaitGroup := r.metadataWaitGroups[entry.Index]
+				if hasWaitGroup {
+					pendingResult = waitGroup
+				} else {
+					pendingResult = &sync.WaitGroup{}
+					pendingResult.Add(1)
+					r.metadataWaitGroups[entry.Index] = pendingResult
+				}
+			}
+		}
+		r.metadataLock.Unlock()
+
+		// The required metadata hasn't been processed yet
+		if pendingResult != nil {
+			pendingResult.Wait()
+			r.metadataLock.Lock()
+			metadata, ok = r.metadataLookup[entry.Index]
+			r.metadataLock.Unlock()
+			// Something went wrong, maybe the reader was closed early
+			if !ok {
+				readResponse.resultErr = errCommitLogReaderMissingLogMetadata
+				outBuf <- readResponse
+				continue
+			}
+		}
 
-	series = metadata
-	datapoint = ts.Datapoint{
-		Timestamp: time.Unix(0, entry.Timestamp),
-		Value:     entry.Value,
+		readResponse.series = metadata
+		readResponse.datapoint = ts.Datapoint{
+			Timestamp: time.Unix(0, entry.Timestamp),
+			Value:     entry.Value,
+		}
+		readResponse.unit = xtime.Unit(byte(entry.Unit))
+		// Copy annotation to prevent reference to pooled byte slice
+		readResponse.annotation = append(readResponse.annotation, entry.Annotation...)
+		// DecRef delayed until this point because even after decoding, underlying
+		// byte slice is still referenced by metadata and annotation
+		arg.bytes.DecRef()
```

Comment:
It seems that we copy the annotation and we also copy the namespace and ID for metadata so we should actually be able to `arg.bytes.Finalize()` here?

Reply:
I wasn't clear about this. Does calling Finalize() mean "I'm done with this, allow it to be re-used"? If so, why do I need to call IncRef() and DecRef() manually?

Reply:
Calling `Finalize()` will return it to the underlying pool.  Inc/Dec is to signify ownership.  When `Finalize()` is called it will make sure the Inc/Dec is zero (no one owns it any longer) or panic.